### PR TITLE
Add a reporting module wrapping standard messages published by nyukis

### PR DESCRIPTION
**Pull request by gdraynz**
First draft that define the global functional perimeter.
Json validation at emission and reception and publication into the `reports` muc, for now.

### CODE REVIEW

File: nyuki/reporting.py [hunk 1]
```diff
@@ -0,0 +1,137 @@
+from datetime import datetime
```
Comment (lerignoux) [inline nyuki/reporting.py]: wouldn't a DB reporting be more adapted (data perenity, ...)

File: nyuki/nyuki.py [hunk 2]
```diff
@@ -102,6 +104,12 @@ def start(self):
         # Start services
         self.loop.run_until_complete(self._services.start())
 
+        if 'bus' in self._services.all:
+            self.reporter = Reporter(self.bus.client.boundjid.user, self.bus)
```
Comment (lerignoux) [inline nyuki/nyuki.py]: wouldn't a DB reporting be more adapted (data perenity, ...)

File: tests/reporting_test.py [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+    def test_001_check_type(self):
+        with assert_raises(ValidationError):
+            self.reporter.check_report({
+                'type': 'something',
+                'author': 'test',
+                'date': datetime.utcnow().isoformat(),
+                'data': {
+                    'address': 'test'
+                }
+            })
```
Comment (lerignoux) [inline tests/reporting_test.py]: you also need to check failing cases

File: nyuki/nyuki.py [hunk 4]
```diff
@@ -180,6 +188,21 @@ def _validate_config(self, config=None):
...
+            log.debug('Report validation failed: %s', ve)
+            return
+        log.info('Valid report received, calling handle_report')
+        await self.handle_report(body)
+
+    async def handle_report(self, body):
+        """
+        Called on reception of a report from another nyuki
+        """
+        log.debug('Report received without callback')
```
Comment (lerignoux) [inline nyuki/nyuki.py]: what od you intend to do with that ?
shouldn't it be a specific nyuki that handles it ? not sure it should be part of nyuki

File: nyuki/reporting.py [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+        'author': {
+            'type': 'string',
+            'minLength': 1
+        },
+        'date': {
+            'type': 'string',
+            'format': 'isoformat'
+        },
+        'data': {'oneOf': [
+            {'$ref': '#/definitions/{}'.format(rtype)} for rtype in ReportTypes.all()
```
Comment (lerignoux) [inline nyuki/reporting.py]: why report type is in data and not type ?

File: nyuki/reporting.py [hunk 6]
```diff
@@ -0,0 +1,137 @@
...
+        return [item.name for item in cls]
+
+
+REPORT_SCHEMA = {
+    'type': 'object',
+    'required': ['type', 'author', 'date', 'data'],
+    'properties': {
+        'type': {
+            'type': 'string',
+            'format': 'report_type'
```
Comment (lerignoux) [inline nyuki/reporting.py]: shouldn't that be a oneOf ?

File: nyuki/nyuki.py (hunk 2 above)
Comment (gdraynz) [inline nyuki/nyuki.py]: It is already tested in the tests below

File: tests/reporting_test.py (hunk 3 above)
Comment (gdraynz) [inline tests/reporting_test.py]: It is already tested in the tests below

File: nyuki/nyuki.py (hunk 4 above)
Comment (gdraynz) [inline nyuki/nyuki.py]: Added this to eventually handle reports from any nyuki, as it is a common functionality to every nyuki

File: nyuki/reporting.py (hunk 5 above)
Comment (gdraynz) [inline nyuki/reporting.py]: `data` contains the report data depending of its own type and so its own schema

File: nyuki/reporting.py (hunk 6 above)
Comment (gdraynz) [inline nyuki/reporting.py]: oneOf needs to be used for definitions, we can't directly check that a string if of an exact value in the schema

File: nyuki/nyuki.py (hunk 2 above)
Comment (gdraynz) [inline nyuki/nyuki.py]: I don't think adding a database in each nyuki is a good thing

Comment (thavel) [inline nyuki/nyuki.py]: I agree with @gdraynz.

File: nyuki/bus.py [hunk 7]
```diff
@@ -316,6 +316,9 @@ def send_message(self, recipient, data):
             return
         if not isinstance(data, dict):
             raise TypeError('Message must be a dict')
+        if not self._connected.is_set():
+            log.info("Waiting for a connection to direct message '%s'", recipient)
+        await self._connected.wait()
```
Comment (thavel) [inline nyuki/bus.py]: Should not be in the `if` condition below?

File: nyuki/nyuki.py [hunk 8]
```diff
@@ -30,8 +31,6 @@ class Nyuki(metaclass=CapabilityHandler):
     single-threaded, asynchronous and concurrent-safe environment.
     The core engine of a nyuki implementation is the asyncio event loop
     (a single loop is used for all features).
-    A wrapper is also provide to ease the use of asynchronous calls
-    over the actions nyukis are inteded to do.
```
Comment (thavel) [inline nyuki/nyuki.py]: :+1:

File: nyuki/nyuki.py [hunk 9]
```diff
@@ -107,6 +107,12 @@ def start(self):
         # Start services
         self.loop.run_until_complete(self._services.start())
 
+        if 'bus' in self._services.all:
+            self.reporter = Reporter(self.bus.client.boundjid.user, self.bus)
+            asyncio.ensure_future(self.bus.subscribe(
+                'reports', self._handle_report
```
Comment (thavel) [inline nyuki/nyuki.py]: Would be great to make the report channel configurable! Default value to `reports` sound good though.

File: nyuki/nyuki.py (hunk 4 above)
Comment (thavel) [inline nyuki/nyuki.py]: I agree with @lerignoux. Let's the nyuki that would handle reports implements its own handlers. See my comment below regarding the channel name.

File: nyuki/reporting.py [hunk 10]
```diff
@@ -0,0 +1,137 @@
...
+                'type': 'string',
+                'minLength': 1
+            }
+        },
+        "additionalProperties": False
+    }
+
+    connection_lost = {
+        'type': 'object',
+        'required': ['address'],
```
Comment (thavel) [inline nyuki/reporting.py]: `address`? At least an `uri`, no?

File: nyuki/reporting.py [hunk 11]
```diff
@@ -0,0 +1,137 @@
...
+            'address': {
+                'type': 'string',
+                'minLength': 1
+            }
+        },
+        "additionalProperties": False
+    }
+
+    @classmethod
+    def all(cls):
```
Comment (thavel) [inline nyuki/reporting.py]: Is it really useful? Could we just code `if something in ReportTypes`?

File: nyuki/reporting.py [hunk 12]
```diff
@@ -0,0 +1,137 @@
...
+
+        try:
+            self.check_report(report)
+        except ValidationError as ve:
+            log.exception(ve)
+            log.error('Report validation failed: %s', str(ve))
+            return
+
+        log.info('Sending report data: %s', report)
+        await self._publisher.publish(report, 'reports')
```
Comment (thavel) [inline nyuki/reporting.py]: See my previous comment on hard coded channel name.

**Comment by thavel** [comment]
On a general level: reports types are strict. `data` content should be more permissive.

File: nyuki/bus.py (hunk 7 above)
Comment (gdraynz) [inline nyuki/bus.py]: It's a deadlock if you `await` it, a call using `ensure_future` will just put it on hold until the connection is set.

File: nyuki/nyuki.py (hunk 9 above)
Comment (gdraynz) [inline nyuki/nyuki.py]: That would means that every call to these reports won't be of any use unless a specific nyuki is up and running.
I figured that any nyuki could listen to these events to eventually do something when, let's say a fellow SMS nyuki loses its connection, another SMS-capable nyuki could hear it and carry on its work...

File: nyuki/reporting.py (hunk 11 above)
Comment (gdraynz) [inline nyuki/reporting.py]: I don't think it works that way :)

File: nyuki/bus.py (hunk 7 above)
Comment (thavel) [inline nyuki/bus.py]: So the task never ends, but the thread isn't blocked, right?

File: nyuki/nyuki.py [hunk 13]
```diff
@@ -89,11 +90,39 @@ def __getattribute__(self, name):
     def config(self):
         return self._config
 
+    @property
+    def report_channel(self):
+        return self.config.get('report_channel', 'errors')
+
+    def _exception_handler(self, loop, context):
+        log.debug('Exception context: %s', context)
+        if 'exception' in context:
```
Comment (thavel) [inline nyuki/nyuki.py]: Erh, few nested `if`/`else` may be replaced by stop conditions:
```
if 'exception' not in context:
    log.warning(context)
    return
```

File: nyuki/nyuki.py [hunk 14]
```diff
@@ -107,6 +136,14 @@ def start(self):
         # Start services
         self.loop.run_until_complete(self._services.start())
 
+        if 'bus' in self._services.all:
+            asyncio.ensure_future(self.bus.subscribe(
```
Comment (thavel) [inline nyuki/nyuki.py]: Still to be removed (aka no default handler), IMO.

File: nyuki/nyuki.py [hunk 15]
```diff
@@ -89,11 +90,39 @@ def __getattribute__(self, name):
...
+                except Exception as e:
+                    log.exception(e)
+                    exc = traceback.format_exc()
+                else:
+                    exc = 'could not retrieve exception'
+            else:
+                log.exception(context['exception'])
+                exc = traceback.format_exc()
+
+            if self.reporter:
```
Comment (thavel) [inline nyuki/nyuki.py]: Do we _really_ want to report all exceptions?

**Pull request merged**